Nfield.Quota/QuotaFrameValidator.cs

```diff
@@ -399,13 +437,126 @@ public QuotaFrameValidator()
+            IEnumerable<QuotaFrameVariable> variables,
+            PropertyValidatorContext context)
+        {
+            bool ProcessLevel(QuotaFrameVariable variable, IEnumerable<QuotaFrameLevel> parents)
+            {
+                // sum of targets, ignoring null values
+                var sum = variable.Levels.Aggregate(0, (total, level) => total + (level.Target ?? 0));
+
+                var isValid = true;
+                foreach (var parent in parents)
+                {
+                    if (parent.MaxTarget != null && parent.MaxTarget < sum)
+                    {
+                        context.MessageFormatter.AppendArgument("VariableName", variable.Name);
+                        context.MessageFormatter.AppendArgument("VariableId", variable.Id);
+                        context.MessageFormatter.AppendArgument("LevelId", parent.Id);
+                        context.MessageFormatter.AppendArgument("LevelName", parent.Name);
+                        context.MessageFormatter.AppendArgument("Sum", sum);
+                        context.MessageFormatter.AppendArgument("MaxTarget", parent.MaxTarget);
+
+                        isValid = false;
+                    }
+                }
+
+                foreach (var level in variable.Levels)
+                {
+                    var newParents = parents.Concat(new[] { level });
+
+                    foreach (var nestedVariable in level.Variables)
+                    {
+                        isValid &= ProcessLevel(nestedVariable, newParents);
+                    }
+                }
+                
+                return isValid;
+            }
+
+            var isTreeValid = true;
+            foreach (var variable in variables)
+            {
+                isTreeValid &= ProcessLevel(variable, new QuotaFrameLevel[0]);
+            }
+
+            return isTreeValid;
+        }
+
+        private static bool HaveNestedMaxLevelsSumToMoreThanMinTargetForEachLevel(
+            QuotaFrame frame,
+            IEnumerable<QuotaFrameVariable> variables,
+            PropertyValidatorContext context)
+        {
+            bool ProcessLevel(QuotaFrameVariable variable, IEnumerable<QuotaFrameLevel> parents)
+            {
+                // we need the sum of the max targets for this variable,
+                // ignoring null values. if *all* values are null, this
+                // validation does not apply (because null means "don't care")
+                var sum = 0;
+                var allTargetsNull = true;
+
+                foreach (var level in variable.Levels)
+                {
+                    if (level.MaxTarget != null)
+                    {
+                        allTargetsNull = false;
+                        sum += level.MaxTarget.Value;
+                    }
+                }
+
+                var isValid = true;
+
+                if (!allTargetsNull)
+                {
+                    foreach (var parent in parents)
```

Comment:
Testing more then one parent level is wrong

Reply:
Why is this wrong? The issue is that you can create an incorrect frame if you leave out targets for variables that are nested but also have children.

Reply:
The wrong is on the foreach.
For more then one parent level up the target can be higher then the sum of the MaxTargets in the levels.